### DISCUSSION TITLE
How to send a text message with link preview?

### DISCUSSION THREAD

**Discussion by david-botelho-mariano** [tulir/whatsmeow#356] [General]
Does anyone have sample code to send text message with preview link?

**Comment by david-botelho-mariano** [comment]
```
//load a file as bytes
thumb, err := ioutil.ReadFile("thumb.jpg")
if err != nil {
	log.Errorf("Failed to load thumb: %v", err)
	return
}		

// construct the message
content := &waProto.Message{
	ExtendedTextMessage: &waProto.ExtendedTextMessage{
		Text:          proto.String("Teste123\n\nhttps://litoranea.fm/?p=21445"),
		Title:         proto.String("Entidades debatem formas de combater o uso do Pix em casos de sequestros relâmpagos - Grupo Litorânea de Comunicação"),
		CanonicalUrl:  proto.String("https://litoranea.fm/entidades-debatem-formas-de-combater-o-uso-do-pix-em-casos-de-sequestros-relampagos"),
		MatchedText:   proto.String("https://litoranea.fm/?p=21445"),
		JpegThumbnail: thumb,
		Description:   proto.String("O Centro de Apoio Operacional Criminal (CAOCrim) do Ministério Público do Estado de São Paulo (MPSP) e promotores de Justiça com atuação na esfera criminal reuniram-se nesta sexta-feira (3), na capital paulista, com representantes da Federação Brasileira de Bancos (Febraban) para discutir formas de combater o uso do Pix como pagamento por resgates em sequestros …"),
	},
}

// parse the JID
recipient, ok := parseJID("120363040918040559@g.us")
if !ok {
	return
}

// send the message
resp, err := cli.SendMessage(context.Background(), recipient, content)
if err != nil {
	log.Errorf("Error sending message: %v", err)
} else {
	log.Infof("Message sent (server timestamp: %s)", resp.Timestamp)
}
```

**Comment by alsak0de** [comment]
Thanks but it did not work for me. Are you positive that the code works? I've tried 'as is' and then adding ThumbnailDirectPath: &resp.DirectPath,	ThumbnailSha256:     resp.FileSHA256, ThumbnailEncSha256:  resp.FileEncSHA256, MediaKey:   resp.MediaKey,

All unsuccessful.

**Reply by david-botelho-mariano** [reply]
I'm sure my code worked because I'm using it in a production environment.

Remember to put the correct link in a similar way to the example and also the "MatchedText" must be inside the "Text" of the "ExtendedTextMessage".

If nothing works, try using my code without any changes so you can see that it works.

**Reply by alsak0de** [reply]
Thanks a lot!!!! Your hint did the trick. I was not having the matchedText into the Text !!!!

**Comment by calliduslabs** [comment]
Hey David. Did you figured out how to show a large thumbnail ?

**Reply by tecbr** [reply]
> Hey David. Did you figured out how to show a large thumbnail ?

I got it with this code:
```
import (
...
"bytes"
"image/jpeg"
)

	targetUrlLower := strings.ToLower(targetUrl)
	if strings.Contains(targetUrlLower, "youtube.com") {
		content.ExtendedTextMessage.PreviewType = waProto.ExtendedTextMessage_VIDEO.Enum()
	} else if strings.HasSuffix(targetUrlLower, ".jpg") || strings.HasSuffix(targetUrlLower, ".jpeg") || strings.HasSuffix(targetUrlLower, ".gif") || strings.HasSuffix(targetUrlLower, ".png") || strings.HasSuffix(targetUrlLower, ".webp") {
		content.ExtendedTextMessage.PreviewType = waProto.ExtendedTextMessage_IMAGE.Enum()
	} else {
		content.ExtendedTextMessage.PreviewType = waProto.ExtendedTextMessage_NONE.Enum()
	}

	// Add thumbnail if available
	if len(thumb) > 0 {
		content.ExtendedTextMessage.JPEGThumbnail = thumb
		resp1, _ := cli.Upload(context.Background(), thumb, "WhatsApp Link Thumbnail Keys")
		content.ExtendedTextMessage.MediaKey = resp1.MediaKey
		content.ExtendedTextMessage.ThumbnailDirectPath = &resp1.DirectPath
		content.ExtendedTextMessage.ThumbnailSHA256 = resp1.FileSHA256
		content.ExtendedTextMessage.ThumbnailEncSHA256 = resp1.FileEncSHA256
		jpeg_config, err := jpeg.DecodeConfig(bytes.NewReader(thumb))
		if err != nil {
			fmt.Printf("error: %v\n", err)
		} else {
			content.ExtendedTextMessage.ThumbnailHeight = proto.Uint32(uint32(jpeg_config.Height))
			content.ExtendedTextMessage.ThumbnailWidth = proto.Uint32(uint32(jpeg_config.Width))
		}
	}
```

**Reply by martinsjr** [reply]
FaviconMMSMetadata ..... perfect

**Comment by calliduslabs** [comment]
Hey @tecbr I didn't tested the code yet but i gave up some time ago with link preview because i mean I got it working but the preview always showed as blurred or small. I ended up using image with caption. Does your code shows a large preview?

**Reply by tecbr** [reply]
Yes. It shows a large preview.

**Reply by calliduslabs** [reply]
By the way, whats your jpeg_config.Height and jpeg_config.Width? I've tested your code it shows a large preview on iphone but on whatsapp web and android it shows a small preview

**Reply by tecbr** [reply]
These are the height and width of the JPEG image. My code was inspired by some discussions:

[Sending a large thumbnail URL in channel messages](https://github.com/tulir/whatsmeow/discussions/498)
[How to send a text message with a large link preview on groups?](https://github.com/tulir/whatsmeow/discussions/765)
[How to send a message with a link to YouTube video, in a way that the video can be played within WhatsApp?](https://github.com/tulir/whatsmeow/discussions/575)
